osu.Game.Tests/Online/TestSceneMultiplayerBeatmapTracker.cs

```diff
@@ -0,0 +1,178 @@
+            addAvailabilityCheckStep("state not downloaded", BeatmapAvailability.NotDownloaded);
+
+            AddStep("undelete beatmap", () => beatmaps.Undelete(testBeatmapSet));
+            addAvailabilityCheckStep("state locally available", BeatmapAvailability.LocallyAvailable);
+        }
+
+        [Test]
+        public void TestTrackerRespectsChecksum()
+        {
+            AddStep("allow importing", () => beatmaps.AllowImport.Set());
+
+            BeatmapInfo wrongBeatmap = null;
+
+            AddStep("import wrong checksum beatmap", () =>
+            {
+                wrongBeatmap = new TestBeatmap(Ruleset.Value).BeatmapInfo;
+                wrongBeatmap.MD5Hash = "1337";
+
+                beatmaps.Import(wrongBeatmap.BeatmapSet).Wait();
+            });
+            AddAssert("wrong beatmap available", () => beatmaps.QueryBeatmap(b => b.OnlineBeatmapID == wrongBeatmap.OnlineBeatmapID) != null);
+            addAvailabilityCheckStep("state still not downloaded", BeatmapAvailability.NotDownloaded);
+
+            AddStep("recreate tracker", () => Child = tracker = new MultiplayerBeatmapTracker
+            {
+                SelectedItem = { BindTarget = selectedItem }
+            });
+            addAvailabilityCheckStep("state not downloaded as well", BeatmapAvailability.NotDownloaded);
+        }
+
+        private void addAvailabilityCheckStep(string description, Func<BeatmapAvailability> expected)
+        {
+            AddAssert(description, () => tracker.Availability.Value.Equals(expected.Invoke()));
+        }
+
+        private string getTestBeatmapOsz()
+        {
+            var filename = Path.GetTempFileName() + ".osz";
+
+            using (var stream = TestResources.OpenResource("Archives/test-beatmap.osz"))
+            using (var file = File.Create(filename))
+                stream.CopyTo(file);
+
+            return filename;
+        }
+
+        private class TestBeatmapManager : BeatmapManager
+        {
+            public readonly ManualResetEventSlim AllowImport = new ManualResetEventSlim();
+
+            protected override ArchiveDownloadRequest<BeatmapSetInfo> CreateDownloadRequest(BeatmapSetInfo set, bool minimiseDownloadSize)
+                => new TestDownloadRequest(set);
+
+            public TestBeatmapManager(Storage storage, IDatabaseContextFactory contextFactory, RulesetStore rulesets, IAPIProvider api, [NotNull] AudioManager audioManager, GameHost host = null, WorkingBeatmap defaultBeatmap = null, bool performOnlineLookups = false)
+                : base(storage, contextFactory, rulesets, api, audioManager, host, defaultBeatmap, performOnlineLookups)
+            {
+            }
+
+            public override async Task<BeatmapSetInfo> Import(BeatmapSetInfo item, ArchiveReader archive = null, CancellationToken cancellationToken = default)
+            {
+                while (!AllowImport.IsSet)
+                    await Task.Delay(10, cancellationToken);
```

Comment:
why does this essentially spin in a loop, instead of properly waiting on the `ManualResetEvent` with a timeout? why use the event at this point if you're going to do that? might as well be a boolean flag?

a `TaskCompletionSource` or a `SemaphoreSlim` is the proper solution for this if you want an async wait.

Reply:
yep, I didn't use `ManualResetEvent.Wait()` because it is not async and I was very unsure about whether it's fine using a synchronous one. `SemaphoreSlim` isn't exactly a "wait for set" type of class, but `TaskCompletionSource` looks to be exactly what I need!

Reply:
`SemaphoreSlim` is applicable. the analogue for setting would be to start with a semaphore with zero count and then call `Release()` when needed.